table/metadata.go

```diff
@@ -0,0 +1,401 @@
+	// LastPartitionSpecID is the highest assigned partition field ID across
+	// all partition specs for the table. This is used to ensure partition
+	// fields are always assigned an unused ID when evolving specs.
+	LastPartitionSpecID() *int
+	// Snapshots returns the list of valid snapshots. Valid snapshots are
+	// snapshots for which all data files exist in the file system. A data
+	// file must not be deleted from the file system until the last snapshot
+	// in which it was listed is garbage collected.
+	Snapshots() []Snapshot
+	// SnapshotByID find and return a specific snapshot by its ID. Returns
+	// nil if the ID is not found in the list of snapshots.
+	SnapshotByID(int64) *Snapshot
+	// SnapshotByName searches the list of snapshots for a snapshot with a given
+	// ref name. Returns nil if there's no ref with this name for a snapshot.
+	SnapshotByName(name string) *Snapshot
+	// CurrentSnapshot returns the table's current snapshot.
+	CurrentSnapshot() *Snapshot
+	// SortOrder returns the table's current sort order, ie: the one with the
+	// ID that matches the default-sort-order-id.
+	SortOrder() SortOrder
+	// SortOrders returns the list of sort orders in the table.
+	SortOrders() []SortOrder
+	// Properties is a string to string map of table properties. This is used
+	// to control settings that affect reading and writing and is not intended
+	// to be used for arbitrary metadata. For example, commit.retry.num-retries
+	// is used to control the number of commit retries.
+	Properties() iceberg.Properties
+}
+
+var (
+	ErrInvalidMetadataFormatVersion = errors.New("invalid or missing format-version in table metadata")
+	ErrInvalidMetadata              = errors.New("invalid metadata")
+)
+
+// ParseMetadata parses json metadata provided by the passed in reader,
+// returning an error if one is encountered.
+func ParseMetadata(r io.Reader) (Metadata, error) {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseMetadataBytes(data)
+}
+
+// ParseMetadataString is like [ParseMetadata], but for a string rather than
+// an io.Reader.
+func ParseMetadataString(s string) (Metadata, error) {
+	return ParseMetadataBytes([]byte(s))
+}
+
+// ParseMetadataBytes is like [ParseMetadataString] but for a byte slice.
+func ParseMetadataBytes(b []byte) (Metadata, error) {
+	ver := struct {
+		FormatVersion int `json:"format-version"`
+	}{}
+	if err := json.Unmarshal(b, &ver); err != nil {
+		return nil, err
+	}
+
+	var ret Metadata
+	switch ver.FormatVersion {
+	case 1:
+		ret = &MetadataV1{}
+	case 2:
+		ret = &MetadataV2{}
+	default:
+		return nil, ErrInvalidMetadataFormatVersion
+	}
+
+	return ret, json.Unmarshal(b, ret)
+}
+
+// https://iceberg.apache.org/spec/#iceberg-table-spec
+type commonMetadata struct {
+	FormatVersion      int                     `json:"format-version"`
+	UUID               uuid.UUID               `json:"table-uuid"`
+	Loc                string                  `json:"location"`
```

Comment:
should it be location instead of Loc ?

Reply:
+1

Reply:
There is already the method `Location()` that returns the location, I can't have a member with the same name as the method. That's why I named it `Loc` originally.